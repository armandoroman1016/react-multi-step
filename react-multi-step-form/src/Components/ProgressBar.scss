.wrapper{
    height: 120px;
}

.progress-bar-container {
    display: flex;
    width: 100%;
    justify-content: space-around;
    align-items: flex-end;
    padding-bottom: 20px;
    
    .progress-step {
        text-align: center;
        width: 25%;
        position: relative;
        .label {
            color:#adadad;
            position: absolute;
            left: 30%;
            width: 46px;
            text-align: center;
            top: 0px;            
        }

        .index {
            height: 30px;
            width: 30px;
            width: 100%;
            display: inline-block;
            transition: background-color 500ms;
            line-height: 20px;
            z-index: 1;
            border-radius: 50%;
            position: absolute;
            left: 30%;
            top: 40px;
            .content {
                border-radius: 50%;
                width: 26px;
                height: 26px;
                padding: 8px;
                border: 2px solid #6cf05b;
                color: #6cf05b;
                text-align: center;
                font: 24px Arial, sans-serif;
                background: linear-gradient(to left,#fff 50%, rgba(#a1fc95, 1) 50%);
                background-position: right bottom;
                background-size: 200% 100%;
            }
            &.completed {
                .content {
                    color: #fff;
                    background-position: left bottom;
                    transition: background 0.3s ease-in;
                }
            }
        }

        .index.completed:after {
            content: "";
            position: absolute;
            left: 2%;
            bottom: 5px;
            height: 3px;
            width: 100%;
            z-index: -1;
            background-color: #a1fc95;
            right: 0;
            transition: all 0.3s ease-in;
            transition-delay: 0.2s;
        }
        .index:after {
            content: "";
            width: 0;
        }

        .index + p{
            font-weight: bold;
        }

        .index.current + p{
            margin: 0;
            padding: 0;
            margin-top: 10px;
            font-size: 18px;
            transition: font-size .25s ease;
            transition-delay: .5s;
        }

        .index.uncomplete + p:only-of-type{
            color: #adadad;
        }
    }
}
