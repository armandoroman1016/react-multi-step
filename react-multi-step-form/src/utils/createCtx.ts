import * as React from "react";

export function createCtx<A extends Record<string, unknown> | null>() {
    const ctx = React.createContext<A | undefined>(undefined);

    function useCtx() {
        const c = React.useContext(ctx);
        if (c === undefined) throw new Error("useCtx muse be inside a Provider with a value");
        return c;
    }

    return [useCtx, ctx.Provider] as const;
}
