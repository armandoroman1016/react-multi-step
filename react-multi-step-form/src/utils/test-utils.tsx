import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
// import { ThemeProvider } from 'my-ui-lib'
// import { TranslationProvider } from 'my-i18n-lib'

import { FormProvider } from "../Contexts/FormContext";

interface Options {
    providerProps?: Record<any, unknown>;
    renderOptions?: RenderOptions;
}

function customRender(ui: ReactElement, options: Options) {
    return render(<FormProvider>{ui}</FormProvider>, options.renderOptions);
}

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
