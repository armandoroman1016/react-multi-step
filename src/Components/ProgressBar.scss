.progress-container{
    border-radius: 2px;
    margin: 10px auto;
    padding: 20px;
    width: 80%;
    display: flex;
    justify-content: space-between;
    position: relative;

}

.progress-step{

    .step-wrapper{
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;
        position: relative;
       
        .step-idx{
            margin: auto 0;
            border-radius: 80%;
            background: #fff;
            border: 1px solid #D3D3D3;
            width: 20px;
            height: 20px;
            padding: 3px;
            text-align: center;
            margin-bottom: 12px;
            color: #D3D3D3;
            &:last-of-type{
                color:red;
            }
            &::before{
                position: absolute;
                z-index: -1;
                top: 22%;
                left: -50%;
                content: "";
                width: 90%;
                height: 1px;
                background: #D3D3D3;
            }

            &.first{
                &::before{
                    width: 0;
                }
            }

            &.completed, &.current{
                background-color: #5c8ef2;
                color: #fff;
                border: none;
                &::before{
                    height: 2px;
                    background: #5c8ef2;
                }
            }

            &.completed{
                font-weight: bold;
            }
        }
    }
}
