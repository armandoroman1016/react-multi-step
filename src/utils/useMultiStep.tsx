import { useContext } from "react";
import { formCtx } from "../contexts/FormContext";

export const useMultiStep = () => {
    const { state, update } = useContext(formCtx);

    function complete(){
        if(!state.errors){
            update({...state, completed: true})
        }
        return state.inputFields
    }

    function updateVals(fieldName: string, fieldValue: string){
        update({ ...state, inputFields: {...state.inputFields, [fieldName]: fieldValue}});
    }

    function setError(b: boolean){
        update({...state, errors: b})
    }

    return { stepForm: state, updateMultiStep: update, complete, updateVals, setError };
};
