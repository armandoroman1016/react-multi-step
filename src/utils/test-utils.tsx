import * as React from "react";
import { render, RenderOptions, RenderResult } from "@testing-library/react";
import { createCtx } from "./createCtx";
import { formCtx, ctx } from "../contexts/FormContext";

export let testCtx: typeof formCtx;
interface Options {
    providerProps?: Record<string, unknown>;
    renderOptions?: RenderOptions;
}

function customRender(ui: React.ReactElement, options: Options): RenderResult {
    const c = { ...ctx, ...options?.providerProps };
    const [formCtx, Provider] = createCtx(c);
    testCtx = formCtx;

    return render(<Provider>{ui}</Provider>, options.renderOptions);
}

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
