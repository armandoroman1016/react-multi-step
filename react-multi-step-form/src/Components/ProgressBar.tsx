import * as React from "react";
import * as CSS from "csstype";

import {useMultiStep} from "../utils/useMultiStep"
interface ProgressBarProps {
    fillColor?: string;
    className?: string;
}

const { useEffect } = React
const ProgressBar = () => {

    const { stepForm } = useMultiStep()

    let steps = 0;

    console.log(stepForm)

    useEffect(() => {
        if(stepForm.maxPosition > 0)steps = stepForm.maxPosition + 1
    }, [stepForm.maxPosition])

    return( 
        <div>
            {stepForm.stepNames && stepForm.stepNames.map(step => {
                return <div>{step}</div>
            }) }
        </div>
    );
};

export default ProgressBar;
