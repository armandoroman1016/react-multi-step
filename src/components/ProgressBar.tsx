import * as React from "react";
import { useMultiStep } from "../utils/useMultiStep";

import styled from "styled-components";

interface ProgressBarProps {
    completedBackground?: string;
    completeFontColor?: string;
    activeBackground?: string;
    activeFontColor?: string;
    incompleteBackground?: string;
    incompleteFontColor?: string;
    errorBackground?: string;
    errorFontColor?: string;
    labelColor?: string;
    font?: string;
    completeLabelColor?: string;
    activeLabelColor?: string;
    errorLabelColor?: string;
    incompleteLabelColor?: string;
}

const ProgressContainer = styled.div`
    border-radius: 2px;
    margin: 10px auto;
    padding: 20px;
    width: 80%;
    display: flex;
    justify-content: space-between;
    position: relative;
`;

const StepItem = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    position: relative;
`;

const StepInfo = styled.div<ProgressBarProps>`
    margin: auto 0;
    border-radius: 80%;
    width: 20px;
    height: 20px;
    padding: 3px;
    text-align: center;
    margin-bottom: 12px;
    background: #fff;
    border: 1px solid #d3d3d3;
    color: #d3d3d3;
    font: ${({ font }) => font || ""};

    &::before {
        position: absolute;
        z-index: -1;
        top: 22%;
        left: -45%;
        content: "";
        width: 90%;
        height: 1px;
        background: #d3d3d3;
    }

    &.first {
        &::before {
            width: 0;
        }
    }

    &.completed {
        background-color: ${({ completedBackground }) => completedBackground || "#5c8ef2"};
        color: ${({ completeFontColor }) => completeFontColor || "#fff"};
        border: 2px solid #fff;
        &::before {
            height: 2px;
            background: ${({ completedBackground }) => completedBackground || "#5c8ef2"};
        }
    }

    &.current {
        background-color: ${({ activeBackground }) => activeBackground || "#5c8ef2"};
        color: ${(props) => {
            return props.activeFontColor || "#fff";
        }};
        border: 2px solid #fff;
        &::before {
            height: 2px;
            background: ${(props) => {
                return props.completedBackground || "#5c8ef2";
            }}};
        }
    }

    &.current.error {
        background-color: ${({ errorBackground }) => errorBackground || "#f73a60"};
        color: ${({ errorFontColor }) => errorFontColor || "#fff"};
    }

    &.completed {
        font-weight: bold;
        cursor: pointer;
    }
`;

const StepLabel = styled.div<ProgressBarProps>`
    font: ${({ font }) => font || ""};

    &.completed {
        color: ${({ completeLabelColor, labelColor }) => completeLabelColor || labelColor || "#5c8ef2"};
    }

    &.current {
        color: ${({ activeLabelColor, labelColor }) => activeLabelColor || labelColor || "#5c8ef2"};
    }

    &.current.error {
        color: ${({ errorLabelColor, errorFontColor }) => errorLabelColor || errorFontColor || "#f73a60"};
    }

    &.uncomplete {
        color: ${({ incompleteLabelColor }) => incompleteLabelColor || "#d3d3d3"};
    }
`;

const ProgressBar = (props: ProgressBarProps) => {
    const { stepForm, updateMultiStep } = useMultiStep();
    const { stepNames, maxPosition, currentPosition, errors } = stepForm;

    const goToStep = (selectedIdx: number) => {
        if (currentPosition > selectedIdx) {
            updateMultiStep({ ...stepForm, currentPosition: selectedIdx, completed: false });
        }
    };

    const getClass = (idx: number) => {
        let str = "";

        if (idx === 0) str += "first ";
        if (currentPosition === idx && errors) return str + "current error";
        if (currentPosition === idx) return str + "current";
        if (currentPosition > idx || currentPosition === maxPosition || stepForm.completed) return str + "completed";
        return str + "uncomplete";
    };

    const getContent = (idx: number) => {
        if (idx < currentPosition || stepForm.completed) return <span>&#10003;</span>;
        return idx + 1;
    };

    return (
        <ProgressContainer>
            {stepNames &&
                stepNames.map((name, idx) => {
                    return (
                        <div key={idx} className="progress-step" style={{ width: `${100 / (maxPosition + 1)}%` }}>
                            <StepItem className="step-wrapper">
                                <StepInfo
                                    activeBackground={props.activeBackground}
                                    activeFontColor={props.activeFontColor}
                                    completeFontColor={props.completeFontColor}
                                    completedBackground={props.completedBackground}
                                    errorBackground={props.errorBackground}
                                    errorFontColor={props.errorFontColor}
                                    incompleteBackground={props.incompleteBackground}
                                    incompleteFontColor={props.incompleteFontColor}
                                    font={props.font}
                                    onClick={() => goToStep(idx)}
                                    className={"step-idx " + getClass(idx)}
                                >
                                    {getContent(idx)}
                                </StepInfo>
                                <StepLabel
                                    className={"label " + getClass(idx)}
                                    font={props.font}
                                    labelColor={props.labelColor}
                                    completeLabelColor={props.completeLabelColor}
                                    activeLabelColor={props.activeLabelColor}
                                    errorLabelColor={props.errorLabelColor}
                                    incompleteLabelColor={props.incompleteLabelColor}
                                    errorFontColor={props.errorFontColor}
                                    activeBackground={props.activeBackground}
                                >
                                    {name}
                                </StepLabel>
                            </StepItem>
                        </div>
                    );
                })}
        </ProgressContainer>
    );
};

export default ProgressBar;
