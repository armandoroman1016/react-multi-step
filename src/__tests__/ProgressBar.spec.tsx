import * as React from "react";
import { render } from "../utils/test-utils";
import ProgressBar from "../components/ProgressBar";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";

afterEach(cleanup);
const stepNames = ["one", "two", "three", "four"];
const maxPosition = 4;
const currentPosition = 0;
const providerProps = { stepNames, currentPosition, maxPosition };

describe("ProgressBar", () => {
    it("renders to the user", async () => {
        const x = render(<ProgressBar />, { providerProps });
        expect(x).toMatchSnapshot();
    });

    it("displays all of the steps", async () => {
        const { queryByText } = render(<ProgressBar />, { providerProps });
        const first = await queryByText(/one/);
        const second = await queryByText(/two/);
        const third = await queryByText(/three/);
        const fourth = await queryByText(/four/);

        expect(first).toBeInTheDocument();
        expect(second).toBeInTheDocument();
        expect(third).toBeInTheDocument();
        expect(fourth).toBeInTheDocument();
    });
    it("displays '✓' on completed steps", async () => {
        const g = render(<ProgressBar />, { providerProps: { ...providerProps, currentPosition: 1 } });
        const x = g.container.querySelector(".step-idx.completed");
        expect(x.textContent).toBe("✓");
    });
    it("will not apply 'error' class when there is no error", () => {
        const { container } = render(<ProgressBar />, { providerProps });
        const x = container.querySelector(".step-idx");
        expect(x.classList.contains("error")).toBe(false);
    });
    it("will apply 'error' class when there is an error", () => {
        const { container } = render(<ProgressBar />, { providerProps: { ...providerProps, errors: true } });
        const x = container.querySelector(".step-idx");
        expect(x.classList.contains("error")).toBe(true);
    });
});
