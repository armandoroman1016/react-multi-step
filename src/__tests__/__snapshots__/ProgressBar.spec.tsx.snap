// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar renders to the user 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ProgressBar__ProgressContainer-sc-15e9afs-0 cZbfYF"
      >
        <div
          class="progress-step"
          style="width: 20%;"
        >
          <div
            class="ProgressBar__StepItem-sc-15e9afs-1 dpFkCC step-wrapper"
          >
            <div
              class="ProgressBar__StepInfo-sc-15e9afs-2 bKMhpD step-idx first current"
            >
              1
            </div>
            <div
              class="ProgressBar__StepLabel-sc-15e9afs-3 hheRIi label first current"
            >
              one
            </div>
          </div>
        </div>
        <div
          class="progress-step"
          style="width: 20%;"
        >
          <div
            class="ProgressBar__StepItem-sc-15e9afs-1 dpFkCC step-wrapper"
          >
            <div
              class="ProgressBar__StepInfo-sc-15e9afs-2 bKMhpD step-idx uncomplete"
            >
              2
            </div>
            <div
              class="ProgressBar__StepLabel-sc-15e9afs-3 hheRIi label uncomplete"
            >
              two
            </div>
          </div>
        </div>
        <div
          class="progress-step"
          style="width: 20%;"
        >
          <div
            class="ProgressBar__StepItem-sc-15e9afs-1 dpFkCC step-wrapper"
          >
            <div
              class="ProgressBar__StepInfo-sc-15e9afs-2 bKMhpD step-idx uncomplete"
            >
              3
            </div>
            <div
              class="ProgressBar__StepLabel-sc-15e9afs-3 hheRIi label uncomplete"
            >
              three
            </div>
          </div>
        </div>
        <div
          class="progress-step"
          style="width: 20%;"
        >
          <div
            class="ProgressBar__StepItem-sc-15e9afs-1 dpFkCC step-wrapper"
          >
            <div
              class="ProgressBar__StepInfo-sc-15e9afs-2 bKMhpD step-idx uncomplete"
            >
              4
            </div>
            <div
              class="ProgressBar__StepLabel-sc-15e9afs-3 hheRIi label uncomplete"
            >
              four
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ProgressBar__ProgressContainer-sc-15e9afs-0 cZbfYF"
    >
      <div
        class="progress-step"
        style="width: 20%;"
      >
        <div
          class="ProgressBar__StepItem-sc-15e9afs-1 dpFkCC step-wrapper"
        >
          <div
            class="ProgressBar__StepInfo-sc-15e9afs-2 bKMhpD step-idx first current"
          >
            1
          </div>
          <div
            class="ProgressBar__StepLabel-sc-15e9afs-3 hheRIi label first current"
          >
            one
          </div>
        </div>
      </div>
      <div
        class="progress-step"
        style="width: 20%;"
      >
        <div
          class="ProgressBar__StepItem-sc-15e9afs-1 dpFkCC step-wrapper"
        >
          <div
            class="ProgressBar__StepInfo-sc-15e9afs-2 bKMhpD step-idx uncomplete"
          >
            2
          </div>
          <div
            class="ProgressBar__StepLabel-sc-15e9afs-3 hheRIi label uncomplete"
          >
            two
          </div>
        </div>
      </div>
      <div
        class="progress-step"
        style="width: 20%;"
      >
        <div
          class="ProgressBar__StepItem-sc-15e9afs-1 dpFkCC step-wrapper"
        >
          <div
            class="ProgressBar__StepInfo-sc-15e9afs-2 bKMhpD step-idx uncomplete"
          >
            3
          </div>
          <div
            class="ProgressBar__StepLabel-sc-15e9afs-3 hheRIi label uncomplete"
          >
            three
          </div>
        </div>
      </div>
      <div
        class="progress-step"
        style="width: 20%;"
      >
        <div
          class="ProgressBar__StepItem-sc-15e9afs-1 dpFkCC step-wrapper"
        >
          <div
            class="ProgressBar__StepInfo-sc-15e9afs-2 bKMhpD step-idx uncomplete"
          >
            4
          </div>
          <div
            class="ProgressBar__StepLabel-sc-15e9afs-3 hheRIi label uncomplete"
          >
            four
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
